### ISSUE TITLE
Decorations won't disappear even when there is no longer a match

### ISSUE THREAD

**Issue by neilsustc** [fabiospampinato/vscode-highlight#3]
My settings

```
"highlight.regexes": {
    "(TODO)": [
        {
            "color": "#C792EA",
            "fontWeight": "bold",
            "overviewRulerColor": "#C792EA",
            "rangeBehavior": 1
        }
    ],
    "(~~.+?~~)": [
        {
            "textDecoration": "line-through"
        }
    ]
}
```

![decoration](https://user-images.githubusercontent.com/7588612/38462647-4b4eeffc-3b1d-11e8-9f62-658546fc24b5.gif)

My guess is you didn't dispose previous decorations on every update

**Comment by fabiospampinato** [comment]
@neilsustc I've seen this happening too. Do you know how am I supposed to dispose of decorations? I can't find anything about this in the docs.

What's weird is that in [Todo+](https://marketplace.visualstudio.com/items?itemName=fabiospampinato.vscode-todo-plus) I'm not doing anything special for clearing out decorations and everything just works 🤔

**Comment by neilsustc** [comment]
Currently, every time `Decorator.decorate()` method is called, you start with a empty `decorations: Map<DecorationType, Range>` and set them to editor.
Instead, you can declare a `Decorator.decorations` field to preserve this map to keep the reference. And then iterate the map calling `DecorationType.dispose()` in the beginning of the method `Decorator.decorate()`

**Comment by fabiospampinato** [comment]
Thanks for the info, I'll fix this tomorrow.

**Comment by fabiospampinato** [comment]
Fixed in v1.0.2.

A little brain dump:

- Disposing of decoration types clears the decorations, but then I think you have to recreate the types if you want to apply new decorations.
- Doing `textEditor.setDecorations ( DecorationType, [] );` clears all of DecorationType's decorations, and it's reasonably fast.
- Performance could be improved by quite a bit if we just did nothing if the decorations types and ranges didn't change from the previous run, but because of `{ rangeBehavior: DecorationRangeBehavior.OpenOpen }` there will be side effects.

**Comment by neilsustc** [comment]
makes sense
